### PR TITLE
ERR: better concat error messages #9157

### LINKED ISSUE #9157: Clearer error message for pd.concat if empty iterator is passed
I just saw following problem on stackoverflow: http://stackoverflow.com/questions/27665560/concatenating-files-with-python-pandas

It is not possible to concatenate an empty sequence of data frames.

This tiny code fragment
```
import pandas as pd
dfs = []
pd.concat(dfs)
``` 

Throws following error:
```
/Users/ch/miniconda/envs/sci33/lib/python3.3/site-packages/pandas/tools/merge.py in __init__(self, objs, axis, join, join_axes, keys, levels, names, ignore_index, verify_integrity, copy)
    765 
    766         if len(objs) == 0:
--> 767             raise ValueError('All objects passed were None')
    768 
    769         # consolidate data & figure out what our result ndim is going to be

ValueError: All objects passed were None
```

I would suggest catching this special case in `pd.concat` before trying to merge and to raise an error with a clearer error description. For this special case, the current error message is not very helpful.

### PR DESCRIPTION
**Pull request by cel4**
closes #9157 

This is a first attempt to improve the error message when trying to concat an empty list/set of objects:

`pd.concat([])` currently raises following error:

```
/Users/ch/miniconda/envs/sci33/lib/python3.3/site-packages/pandas/tools/merge.py in __init__(self, objs, axis, join, join_axes, keys, levels, names, ignore_index, verify_integrity, copy)
    765 
    766         if len(objs) == 0:
--> 767             raise ValueError('All objects passed were None')
    768 
    769         # consolidate data & figure out what our result ndim is going to be

ValueError: All objects passed were None
```

After the tiny fix it is:

```
/Users/ch/repo/pandas/pandas/tools/merge.py in __init__(self, objs, axis, join, join_axes, keys, levels, names, ignore_index, verify_integrity, copy)
    767 
    768         if len(objs) == 0:
--> 769             raise ValueError('No objects passed')
    770 
    771         if isinstance(objs, dict):

ValueError: No objects passed
```

### CODE REVIEW

File: pandas/tools/merge.py [hunk 1]
```diff
@@ -769,9 +770,16 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
             if keys is None:
                 keys = sorted(objs)
             objs = [objs[k] for k in keys]
+            obj_count = len(objs)
+        else:
+            count_gen, objs = itertools.tee(objs)
+            obj_count = len(list(count_gen))
+
+        if obj_count == 0:
```
Comment (jreback) [inline pandas/tools/merge.py]: doesn't ``len(objects)`` accomplish this?

**Comment by jreback** [comment]
needs a test

File: pandas/tools/merge.py (hunk 1 above)
Comment (cel4) [inline pandas/tools/merge.py]: Well, it does look a little hackish. I had issues with `objs` being a generator. `len(objs)` would consume the generator, that's why I used the hack with `itertools.tee`.

**Comment by cel4** [comment]
Yes, still thinking how to design a good test. I still raise a ValueError, but this time with a slightly different error message. Should I test for the wording of the error message? I am not too happy with this solution since it breaks if someone decides to come up with a better message.

**Comment by jreback** [comment]
@cel4 you can do a ``assertRegExp``. And it *should* break if something (e.g. the error message is changed), that's the point.

**Comment by cel4** [comment]
@jreback, sorry I am very busy at the moment. I am planning to continue the work here in a week or two.

**Comment by cel4** [comment]
I added a draft version how a test could look like. Would appreciate any comments.

**Comment by cel4** [comment]
`AttributeError: 'TestOrderedMerge' object has no attribute 'assertRegexpMatches'` - grr...  seems like `assertRegexpMatches` is not part of the `unittest` module of the `2.6` version...

**Comment by jorisvandenbossche** [comment]
@cel4 In the pandas `testing` module, there is a `assertRaisesRegexp` available for this -> `tm.assertRaisesRegexp()`

**Comment by cel4** [comment]
Hmh, tests are green. Yet I am not sure if I like this solution. Wouldn't it be better if we also check in`pd.concat` for empty sequences?

**Comment by jreback** [comment]
this needs to address my comments above
you can force materializing a generator and thrn just take the len

**Comment by sergeny** [comment]
I propose "ValueError: No objects to concatenate"

**Comment by cel4** [comment]
@sergeny, I like that suggestion. Unfortunately, I do not have much time at the moment and I totally forgot about that PR. I hope I can finish it soon.

**Comment by jreback** [comment]
closing as stale, but if you would like to reopen and fixup pls do.

**Comment by cel4** [comment]
Should be ready for a first round of reviews.

File: pandas/tools/tests/test_merge.py [hunk 2]
```diff
@@ -2552,6 +2552,22 @@ def _constructor(self):
 
         tm.assertIsInstance(result, NotADataFrame)
 
+    def test_empty_sequence_concat(self):
+        empty_pat = "[Nn]o objects"
```
Comment (jreback) [inline pandas/tools/tests/test_merge.py]: add the issue number here as a comment

**Comment by jreback** [comment]
looks reasonable. pls add a release note. and squash.

**Comment by cel4** [comment]
When travis is happy, this should be ready for review once again.

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/58db03eec9bd6d78976b844267c3c1a9f61a8eb3

thanks!

**Pull request closed without merging**